boards/beagle/beagleconnect_freedom/board_antenna.c

```diff
@@ -1,116 +1,119 @@
-/* SPDX-License-Identifier: Apache-2.0
- *
+/*
  * Copyright (c) 2021 Florin Stancu
- * Copyright (c) 2021 Jason Kridner, BeagleBoard.org Foundation
+ * Copyright (c) 2024 Ayush Singh <ayush@beagleboard.org>
  *
+ * SPDX-License-Identifier: Apache-2.0
  */
 
 /*
  * Implements the RF driver callback to configure the on-board antenna
  * switch.
  */
 
-#include <zephyr/kernel.h>
-#include <zephyr/device.h>
+#define DT_DRV_COMPAT skyworks_sky13317
+
 #include <zephyr/init.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/pinctrl.h>
 
 #include <ti/drivers/rf/RF.h>
-#include <driverlib/gpio.h>
-#include <driverlib/ioc.h>
 #include <driverlib/rom.h>
+#include <driverlib/interrupt.h>
 
-/* DIOs for RF antenna paths */
-#define BOARD_RF_HIGH_PA   29		/* TODO: pull from DT */
-#define BOARD_RF_SUB1GHZ   30		/* TODO: pull from DT */
+/* custom pinctrl states for the antenna mux */
+#define PINCTRL_STATE_ANT_SUBG    1
+#define PINCTRL_STATE_ANT_SUBG_PA 2
 
-static void board_cc13xx_rf_callback(RF_Handle client, RF_GlobalEvent events,
-		void *arg);
+#define BOARD_ANT_GPIO_PA   0
+#define BOARD_ANT_GPIO_SUBG 1
 
+static int board_antenna_init(const struct device *dev);
+static void board_cc13xx_rf_callback(RF_Handle client, RF_GlobalEvent events, void *arg);
 
 const RFCC26XX_HWAttrsV2 RFCC26XX_hwAttrs = {
-	.hwiPriority        = INT_PRI_LEVEL7,
-	.swiPriority        = 0,
+	.hwiPriority = INT_PRI_LEVEL7,
+	.swiPriority = 0,
 	.xoscHfAlwaysNeeded = true,
 	/* RF driver callback for custom antenna switching */
 	.globalCallback = board_cc13xx_rf_callback,
 	/* Subscribe to events */
-	.globalEventMask = (RF_GlobalEventRadioSetup |
-			RF_GlobalEventRadioPowerDown),
+	.globalEventMask = (RF_GlobalEventRadioSetup | RF_GlobalEventRadioPowerDown),
 };
 
+PINCTRL_DT_INST_DEFINE(0);
+DEVICE_DT_INST_DEFINE(0, board_antenna_init, NULL, NULL, NULL, POST_KERNEL,
+		      CONFIG_BOARD_ANTENNA_INIT_PRIO, NULL);
+
+#define GPIO_DT_SPEC(n, p, i) GPIO_DT_SPEC_GET_BY_IDX(n, p, i),
+
+static const struct pinctrl_dev_config *ant_pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(0);
+static const struct gpio_dt_spec ant_gpios[] = {
+	DT_FOREACH_PROP_ELEM(DT_NODELABEL(antenna_mux0), gpios, GPIO_DT_SPEC)};
+
 /**
  * Antenna switch GPIO init routine.
  */
-static int board_antenna_init(void)
+int board_antenna_init(const struct device *dev)
 {
+	ARG_UNUSED(dev);
+	int i;
 
-	/* set all paths to low */
-	IOCPinTypeGpioOutput(BOARD_RF_HIGH_PA);
-	GPIO_setOutputEnableDio(BOARD_RF_HIGH_PA, GPIO_OUTPUT_DISABLE);
-	IOCPinTypeGpioOutput(BOARD_RF_SUB1GHZ);
-	GPIO_setOutputEnableDio(BOARD_RF_SUB1GHZ, GPIO_OUTPUT_DISABLE);
+	/* default pinctrl configuration: set all antenna mux control pins as GPIOs */
+	pinctrl_apply_state(ant_pcfg, PINCTRL_STATE_DEFAULT);
+	/* set all GPIOs to 0 (all RF paths disabled) */
+	for (i = 0; i < ARRAY_SIZE(ant_gpios); i++) {
+		gpio_pin_configure_dt(&ant_gpios[i], 0);
+	}
```

Comment:
I'm a bit confused on the approach. You have states to toggle between defined in the DT, but those states are only associated with the pinmux. That seems elegant for the pinmux, but now you are manually poking the GPIO state itself. Why not define both in DT if that is where you are moving things?

Reply:
It seems some manual poking is required due to the bug in communication between cc1352r and cc1352p variants making more poking like this necessary. This was present in out old version and is still present in the launchxl board antenna. I am just following what Ti is doing here. Maybe @vaishnavachath, @niflostancu and @stephanosio can provide more insight

Reply:
@vaishnavachath any comment?

Let's not follow blindly, but rather use the DT settings provided for pinmux and GPIO to set the appropriate mode. Adding mode-specific logic defeats the point of using DT to provide the mode settings and removes the flexibility. The specific modes should all be resolved statically.

Reply:
So, I did go through the technical spec and it states that manual poking should not really be required. However, it also uses `RFC_GPO1` which I guess doesn't work great according to the comment below:
```
/* Note: RFC_GPO3 is a work-around because the RFC_GPO1 */
/* is sometimes not de-asserted on CC1352 Rev A. */
```
Thus, I am not confident in removing stuff that might be present due to some hardware issues that are not documented in their Technical Spec. At the same time, Beagleplay cc1352 support has already missed the deadline for 3.7 due to this blocker so I would really like to move things along.

I can remove the GPIO and check if bcf to bcf works but I cannot guarantee communication with any other subg hardware if we deviate from launchxl too much.

Reply:
@Ayush1325, you're right, I already stumbled over the same problem. It should have been documented better in the original file and in TI's errata, IMO. But that's unrelated to this PR. Resolving.